### ISSUE TITLE
Scoped dependencies for messages initiated on the server side?

### ISSUE THREAD

**Issue by roman-bilyi** [graphql-dotnet/graphql-dotnet#3546]
## Summary
Do we have a DI **scope** bound to the web socket so we can resolve dependencies for sending server messages into the subscription using _that scope_? 

## Relevant information
I have some dependencies that rely on HttpContext so we can get things like "host", "currentUser" etc. Those are resolved ok on the subscription requests. Understandably, that we can't have a HttpContext when we fire some events from the server. 
I hoped that IResolveFieldContext.RequestServices are resolved in the scope of subscription requests and we can safely get, for example, `CurrentUser`. 
For sure, relying on the `HttpContext` may be wrong in the first place, but I'm just trying to describe the whole picture and the reasons I'm facing.
I'm sure there's a way. I'd like to ask, how to solve it correctly.

### Environment (if relevant)
GraphQL 7.2.2
.NET 6

**Comment by Shane32** [comment]
Yes ... and no.

What's important here is that the `HttpContextAccessor` instance is a singleton, and is not bound by the service scope, but by an instance of `AsyncLocal` ([see here for details](https://learn.microsoft.com/en-us/dotnet/api/system.threading.asynclocal-1?view=net-7.0)).  So for that specific scenario, you should be asking if the resolver executes within the async context/stack of the HTTP request, as you can just inject `IHttpContextAccessor` via DI into the graph type.  I'm sorry, I'm not sure offhand but I expect so.  I'll probably write a test for this and commit it to the repo is it is known for sure.

If you want to execute scoped services within a subscription, please see [this link](https://github.com/graphql-dotnet/server#service-scope).  In short:

```cs
services.AddGraphQL(b => b
    .AddAutoSchema<Query>()
    .AddSystemTextJson()
    // configure queries for serial execution (optional)
    .AddExecutionStrategy<SerialExecutionStrategy>(OperationType.Query)
    // configure subscription field resolvers for scoped serial execution (parallel is optional)
    .AddScopedSubscriptionExecutionStrategy());
```

Generally when you need to use scoped services, this implies serial execution also, since almost never are scoped services designed for simultaneous multithreaded access.

On a side note, `IResolveFieldContext.User` contains the `ClaimsPrincipal` of the user, copied from the `HttpContext.User` property.  You can also use the [user context builder](https://github.com/graphql-dotnet/server#user-context-builder) to create a custom user context containing any data you wish to copy from the `HttpContext`.  The user context will be passed through to all resolvers.

**Comment by Shane32** [comment]
See: https://github.com/dotnet/aspnetcore/blob/main/src/Http/Http/src/HttpContextAccessor.cs

**Comment by Shane32** [comment]
I have verified that GraphQL field resolvers during subscription data events occurs within the asynchronous context of the sender of the data event, **not the HTTP context**.  This should be treated as a bug and fixed within SubscriptionExecutionStrategy.

See:
- https://github.com/graphql-dotnet/server/pull/991

**Comment by Shane32** [comment]
I'm classifying this as a bug, because any GraphQL field resolvers would expect to run in the context of the client (and so access the client's `HttpContext` instance) rather than the event source, which could be another user.  This could potentially be a security flaw if any authentication code inspects `HttpContext.User` rather than `IResolveFieldContext.User` within a field resolver.

**Comment by Shane32** [comment]
@roman-bilyi For an immediate fix, you'd need to duplicate SubscriptionExecutionStrategy.cs and ObservableExtensions.cs, patching the latter with the fix shown in #3547 , and use that execution strategy for subscriptions.  If you use `ScopedSubscriptionExecutionStrategy`, you'd need to duplicate that as well so the patched `SubscriptionExecutionStrategy` is referenced.

Your patched execution strategy would be installed via `.AddExecutionStrategy<MySubscriptionExecutionStrategy>(OperationType.Subscription)` within your `AddGraphQL` call.

**Comment by roman-bilyi** [comment]
@Shane32 Thank you A LOT for the help and very detailed explanations :-) I'll go with the solution you've provided.
